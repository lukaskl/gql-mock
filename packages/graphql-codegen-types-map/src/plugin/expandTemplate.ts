import * as changeCase from 'change-case'
import { parse } from 'esprima'
import { OperationTypeNode } from 'graphql'
import evaluate from 'static-eval'

export interface TemplateVariables {
  operationName: string
  operationKind: OperationTypeNode
}

const casingOperations = (Object.keys(changeCase) as (keyof typeof changeCase)[])
  .map(x => ({ [x]: changeCase[x] }))
  .reduce((l, r) => ({ ...l, ...r }), {})

export const expandTemplate = (template: string, variables: TemplateVariables) => {
  const ast = (parse('`' + template.replace(/{/g, '${') + '`').body[0] as any).expression

  const result: string = evaluate(ast, {
    ...casingOperations,
    ...variables,
  })
  if (result.includes('[object Object]')) {
    const errorMessage =
      `Invalid variable or function name used in "operationsMap" template. ` +
      `Given template: "${template}". ` +
      `Allowed variables: "${Object.keys(variables).join(', ')}". ` +
      `Allowed functions: "${Object.keys(casingOperations).join(', ')}". ` +
      `e.g. template: "{pascalCase(operationName)}{pascalCase(operationKind)}Variables"`
    console.error(`ERROR: ${errorMessage}`)
    throw new Error(errorMessage)
  }
  return result
}

export const validateTemplates = (...templates: string[]) => {
  const validationVariables: TemplateVariables = {
    operationKind: 'query',
    operationName: 'validation',
  }
  for (const template of templates) {
    expandTemplate(template, validationVariables)
  }
}
