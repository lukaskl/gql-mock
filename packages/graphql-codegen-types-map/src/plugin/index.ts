export { plugin } from './plugin'
export { OperationsParser } from './parser'
