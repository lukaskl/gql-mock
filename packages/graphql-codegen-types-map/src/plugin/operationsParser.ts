import {
  ASTNode,
  DocumentNode,
  VariableDefinitionNode,
  visit,
  OperationTypeNode,
  print,
  TypeInfo,
  visitWithTypeInfo,
  BREAK,
} from 'graphql'

export interface OperationsMapEntry {
  operationName: string
  hasVariables: boolean
  operationKind: OperationTypeNode
}

export const getVariablesAst = (node: ASTNode) => {
  let variable: VariableDefinitionNode | undefined = undefined
  visit(node, {
    VariableDefinition: node => {
      variable = node
      return BREAK
    },
  })

  return variable
}

export const getOperationsMap = (typeInfo: TypeInfo, allAst: DocumentNode) => {
  const operations: OperationsMapEntry[] = []
  visit(
    allAst,
    visitWithTypeInfo(typeInfo, {
      leave: {
        OperationDefinition: node => {
          const name = node.name?.value
          if (!name) {
            throw new Error(`Anonymous operations are not allowed: ${print(node)}`)
          }
          const operationKind = node?.operation
          const variables = getVariablesAst(node)
          operations.push({
            operationName: name,
            hasVariables: !!variables,
            operationKind,
          })
        },
      },
    })
  )
  return operations
}
