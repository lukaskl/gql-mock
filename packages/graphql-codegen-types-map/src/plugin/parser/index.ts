export { OperationsParser } from './OperationsParser'
export { FragmentUsage, TypeUsages, getTypeUsages, UsagePathNode } from './getTypeUsages'
