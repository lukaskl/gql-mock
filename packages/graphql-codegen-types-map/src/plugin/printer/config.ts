import {
  validateFieldArgsTemplate,
  validateOperationsTemplate,
  validateTypeAccessorTemplate,
} from './expandTemplate'
import merge from 'lodash.merge'
import { pickKeys, KeysObj } from '~/utils'

export interface OperationsTemplatesConfig {
  operationTypeTemplate: string
  operationDocumentTemplate: string
  variablesTypeTemplate: string
  operationKindTemplate: string
  typeUsagesTemplate: string
}
export interface FieldArgsTemplateConfig {
  fieldArgsTypeTemplate: string
}

export interface TypeAccessorTemplateConfig {
  typeAccessorTypeTemplate: string
}

export type TemplatesConfig = OperationsTemplatesConfig & FieldArgsTemplateConfig & TypeAccessorTemplateConfig

export interface TypesImportConfig {
  importTypesFrom: string | undefined
  importedTypesAlias: string
}

export interface TypeUsagesConfig extends Pick<OperationsTemplatesConfig, 'typeUsagesTemplate'> {
  withTypeUsages: boolean
}

export interface AllConfigOptions extends TemplatesConfig, TypesImportConfig, TypeUsagesConfig {
  withDocumentsMap: boolean
}

export const defaultConfig: AllConfigOptions = {
  operationTypeTemplate: '{OperationName}{OperationKind}',
  variablesTypeTemplate: '{OperationName}{OperationKind}Variables',
  operationDocumentTemplate: '{OperationName}{OperationKind === "Fragment" ? "FragmentDoc" : "Document"}',
  operationKindTemplate: '{operationKind}',
  typeAccessorTypeTemplate: '{typeKind === "scalar" ? `Scalars["{typeName}"]` : TypeName }',
  importedTypesAlias: 'Types',
  importTypesFrom: undefined,
  typeUsagesTemplate: 'TypeUsagesFor{OperationName}{OperationKind}',
  fieldArgsTypeTemplate: '{ParentName}{FieldName}Args',
  withTypeUsages: true,
  withDocumentsMap: true,
}

const operationsTemplatesConfigKeysObs: KeysObj<keyof OperationsTemplatesConfig> = {
  operationKindTemplate: null,
  operationTypeTemplate: null,
  variablesTypeTemplate: null,
  typeUsagesTemplate: null,
  operationDocumentTemplate: null,
}
const fieldArgsTemplatesConfigKeysObs: KeysObj<keyof FieldArgsTemplateConfig> = {
  fieldArgsTypeTemplate: null,
}
const typeAccessorTemplatesConfigKeysObs: KeysObj<keyof TypeAccessorTemplateConfig> = {
  typeAccessorTypeTemplate: null,
}
const typesImportConfigKeysObs: KeysObj<keyof TypesImportConfig> = {
  importTypesFrom: null,
  importedTypesAlias: null,
}
const typeUsagesConfigKeysObs: KeysObj<keyof TypeUsagesConfig> = {
  withTypeUsages: null,
  typeUsagesTemplate: null,
}

export class PrinterConfig {
  private readonly config: AllConfigOptions

  constructor(rawConfig: Partial<AllConfigOptions> = {}) {
    this.config = merge(defaultConfig, rawConfig)
  }

  get allOptions() {
    return this.config
  }

  get operationTemplates() {
    return this.pickConfigEntries(operationsTemplatesConfigKeysObs)
  }

  get fieldArgsTemplates() {
    return this.pickConfigEntries(fieldArgsTemplatesConfigKeysObs)
  }

  get typeAccessorTemplates() {
    return this.pickConfigEntries(typeAccessorTemplatesConfigKeysObs)
  }

  get importTypes() {
    const config = this.pickConfigEntries(typesImportConfigKeysObs)
    const importRef = config.importTypesFrom ? `${config.importedTypesAlias}.` : ''
    return { ...config, importRef }
  }

  get typeUsages() {
    return this.pickConfigEntries(typeUsagesConfigKeysObs)
  }

  validateConfig() {
    validateOperationsTemplate(...Object.values(this.operationTemplates))
    validateFieldArgsTemplate(...Object.values(this.fieldArgsTemplates))
    validateTypeAccessorTemplate(...Object.values(this.typeAccessorTemplates))
  }

  private pickConfigEntries = <T extends keyof AllConfigOptions>(pick: KeysObj<T>) =>
    pickKeys<T, AllConfigOptions>(this.config, pick)
}
