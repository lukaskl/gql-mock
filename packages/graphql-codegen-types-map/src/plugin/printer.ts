import { expandTemplate } from './expandTemplate'
import { OperationsMapEntry } from './operationsParser'
import { TemplatesConfig } from './plugin'

export interface OperationsMapPrinterConfig {
  operationsMap?: {
    operationKindTemplate?: string
    operationTypeTemplate?: string
    variablesTypeTemplate?: string
  }
}

export class OperationsPrinter {
  constructor(protected readonly operations: OperationsMapEntry[], readonly templates: TemplatesConfig) {}

  get operationsMapFields() {
    const { operationKindTemplate, operationTypeTemplate, variablesTypeTemplate } = this.templates

    return this.operations
      .map(x => {
        const opKind = expandTemplate(operationKindTemplate, x)
        const opType = expandTemplate(operationTypeTemplate, x)
        const varType = expandTemplate(variablesTypeTemplate, x)

        return `'${x.operationName}': { operationType: ${opType}, variablesType: ${varType}, kind: '${opKind}' }`
      })
      .join('\n  ')
  }

  get operationsMapInterface() {
    return `export interface OperationsMap {\n  ${this.operationsMapFields}\n}`
  }
}
