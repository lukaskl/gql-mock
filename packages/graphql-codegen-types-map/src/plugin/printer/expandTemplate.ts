import * as changeCase from 'change-case'
import { parse } from 'esprima'
import { OperationTypeNode } from 'graphql'
import evaluate from 'static-eval'
import { ERRORS, fromEntries } from '~/utils'
import { GraphQLSchemaOutputTypes } from '../parser'

const casingOperations = (Object.keys(changeCase) as (keyof typeof changeCase)[])
  .map(x => ({ [x]: changeCase[x] }))
  .reduce((l, r) => ({ ...l, ...r }), {})

export type OperationKind = OperationTypeNode | 'fragment'
export interface OperationTemplateVariables {
  operationName: string
  operationKind: OperationKind
}

export interface FieldArgsTemplateVariables {
  parentName: string
  fieldName: string
}

export interface TypeAccessorTemplateVariables {
  typeName: string
  typeKind: GraphQLSchemaOutputTypes
}

export type TemplateVariables =
  | FieldArgsTemplateVariables
  | OperationTemplateVariables
  | TypeAccessorTemplateVariables

export const expandTemplate = <T extends TemplateVariables>(template: string, variables: T): string => {
  const ast = (parse('`' + template.replace(/{/g, '${') + '`').body[0] as any).expression

  const allVariables = {
    ...fromEntries(
      Object.entries(variables).map(([name, value]) => [
        changeCase.pascalCase(name),
        changeCase.pascalCase(value),
      ])
    ),
    ...variables,
  }

  const result: string = evaluate(ast, {
    ...casingOperations,
    ...allVariables,
  })
  if (result.includes('[object Object]')) {
    return ERRORS.invalidTemplate(template, allVariables, casingOperations)
  }
  return result
}

const validateTemplate = (templates: string[], dummyVariables: TemplateVariables) => {
  for (const template of templates) {
    expandTemplate(template, dummyVariables)
  }
}

export const validateOperationsTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    operationKind: 'query',
    operationName: 'validation',
  })

export const validateFieldArgsTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    fieldName: 'field-name',
    parentName: 'parent-name',
  })

export const validateTypeAccessorTemplate = (...templates: string[]) =>
  validateTemplate(templates, {
    typeKind: 'interface',
    typeName: 'fake-name',
  })
