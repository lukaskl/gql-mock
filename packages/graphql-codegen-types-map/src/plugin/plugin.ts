import { PluginFunction } from '@graphql-codegen/plugin-helpers'
import { concatAST, DocumentNode, TypeInfo } from 'graphql'

import { validateTemplates } from './expandTemplate'
import { getOperationsMap } from './operationsParser'
import { OperationsMapPrinterConfig, OperationsPrinter } from './printer'

const ensureConfigDefaults = (config: OperationsMapPrinterConfig) => {
  const {
    operationsMap: {
      operationTypeTemplate = '{pascalCase(operationName)}{pascalCase(operationKind)}',
      variablesTypeTemplate = '{pascalCase(operationName)}{pascalCase(operationKind)}Variables',
      operationKindTemplate = '{operationKind}',
    } = {},
  } = config

  validateTemplates(operationTypeTemplate, variablesTypeTemplate, operationKindTemplate)
  return { operationTypeTemplate, variablesTypeTemplate, operationKindTemplate }
}

export type TemplatesConfig = ReturnType<typeof ensureConfigDefaults>

export const plugin: PluginFunction<OperationsMapPrinterConfig> = (schema, rawDocuments, config) => {
  const templates = ensureConfigDefaults(config)
  const typeInfo = new TypeInfo(schema)

  const documents = rawDocuments
  const allAst = concatAST(
    documents
      .map(
        v =>
          v.document ||
          // The next line is to make it work with older versions (<= v1.6.1) of @graphql-codegen
          (v as any).content
      )
      .filter(x => !!x) as DocumentNode[]
  )

  const operations = getOperationsMap(typeInfo, allAst)
  const operationsMapPrinter = new OperationsPrinter(operations, templates)

  return operationsMapPrinter.operationsMapInterface
}
