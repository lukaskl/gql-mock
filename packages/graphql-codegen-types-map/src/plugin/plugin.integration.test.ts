import { gitHuntSchema } from '~/test-support'
import { plugin } from './plugin'
import { AllConfigOptions } from './printer'

const runPlugin = async (config: Partial<AllConfigOptions> = {}) => {
  return await plugin(
    gitHuntSchema.schema,
    gitHuntSchema.documents,
    { operationsMap: config },
    {
      outputFile: '',
    }
  )
}

describe('TypeScript Operations Map Plugin', () => {
  it('Test 1', async () => {
    const result = await runPlugin()

    expect(result).toMatchSnapshot()
  })
})
