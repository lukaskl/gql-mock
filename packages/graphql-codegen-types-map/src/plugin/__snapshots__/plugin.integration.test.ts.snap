// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Operations Map Plugin Test 1 1`] = `
Object {
  "append": Array [],
  "content": "

export type AllOutputTypes = {
  'Query': Query
  'Followable': Followable
  'Organization': Organization
  'Actor': Actor
  'Repository': Repository
  'FeedType': FeedType
  'Entry': Entry
  'User': User
  'Comment': Comment
  'Vote': Vote
  'Mutation': Mutation
  'VoteType': VoteType
  'Subscription': Subscription
}

export type AllScalarTypes = {
  'String': Scalars[\\"String\\"]
  'Int': Scalars[\\"Int\\"]
  'Date': Scalars[\\"Date\\"]
  'Float': Scalars[\\"Float\\"]
  'Boolean': Scalars[\\"Boolean\\"]
}

export interface FieldArgsUsagesMap {
  'Query': {
    feed: QueryFeedArgs,
    entry: QueryEntryArgs,
    repository: QueryRepositoryArgs
  },
  'Entry': {
    comments: EntryCommentsArgs
  },
  'Mutation': {
    submitRepository: MutationSubmitRepositoryArgs,
    vote: MutationVoteArgs,
    submitComment: MutationSubmitCommentArgs
  },
  'Subscription': {
    commentAdded: SubscriptionCommentAddedArgs
  }
}

export type TypeUsagesForCommentsPageCommentFragment = { User: CommentsPageCommentFragment['postedBy'] } &
  { Comment: CommentsPageCommentFragment }

  export type TypeUsagesForFeedEntryFragment = { Repository: FeedEntryFragment['repository'] } &
  { Actor: FeedEntryFragment['repository']['owner'] } &
  { Entry: FeedEntryFragment } &
  TypeUsagesForVoteButtonsFragment &
  TypeUsagesForRepoInfoFragment

  export type TypeUsagesForRepoInfoFragment = { Repository: RepoInfoFragment['repository'] } &
  { User: RepoInfoFragment['postedBy'] } &
  { Entry: RepoInfoFragment }

  export type TypeUsagesForVoteButtonsFragment = { Vote: VoteButtonsFragment['vote'] } &
  { Entry: VoteButtonsFragment }

  export type TypeUsagesForOnCommentAddedSubscription = { Comment: OnCommentAddedSubscription['commentAdded'] } &
  { User: NN<OnCommentAddedSubscription['commentAdded']>['postedBy'] } &
  { Subscription: OnCommentAddedSubscription }

  export type TypeUsagesForCommentQuery = { User: CommentQuery['currentUser'] } &
  { Entry: CommentQuery['entry'] } &
  { User: NN<CommentQuery['entry']>['postedBy'] } &
  { Comment: NN<CommentQuery['entry']>['comments'][0] } &
  { Comment: NN<CommentQuery['entry']>['comments2'][0] } &
  { Comment: NN<CommentQuery['entry']>['comments3'][0] } &
  { Comment: UU<NN<NN<CommentQuery['entry']>['comments3'][0]>, 'Comment'> } &
  { Repository: NN<CommentQuery['entry']>['repository'] } &
  { Repository: UU<NN<CommentQuery['entry']>['repository'], 'Repository'> } &
  { Query: CommentQuery } &
  TypeUsagesForCommentsPageCommentFragment

  export type TypeUsagesForCurrentUserForProfileQuery = { User: CurrentUserForProfileQuery['currentUser'] } &
  { Query: CurrentUserForProfileQuery }

  export type TypeUsagesForFeedQuery = { User: FeedQuery['currentUser'] } &
  { Entry: NN<FeedQuery['feed']>[0] } &
  { Query: FeedQuery } &
  TypeUsagesForFeedEntryFragment

  export type TypeUsagesForGetExtendedFollowSuggestionsQuery = { Followable: GetExtendedFollowSuggestionsQuery['followSuggestion'] } &
  { Organization: UU<NN<GetExtendedFollowSuggestionsQuery['followSuggestion']>, 'Organization'> } &
  { Repository: UU<NN<GetExtendedFollowSuggestionsQuery['followSuggestion']>, 'Repository'> } &
  { Actor: NN<UU<NN<GetExtendedFollowSuggestionsQuery['followSuggestion']>, 'Repository'>['contributors']>[0] } &
  { User: UU<NN<NN<UU<NN<GetExtendedFollowSuggestionsQuery['followSuggestion']>, 'Repository'>['contributors']>[0]>, 'User'> } &
  { Organization: UU<NN<NN<UU<NN<GetExtendedFollowSuggestionsQuery['followSuggestion']>, 'Repository'>['contributors']>[0]>, 'Organization'> } &
  { Query: GetExtendedFollowSuggestionsQuery }

  export type TypeUsagesForGetSimpleFollowSuggestionsQuery = { Followable: GetSimpleFollowSuggestionsQuery['followSuggestion'] } &
  { Organization: UU<NN<GetSimpleFollowSuggestionsQuery['followSuggestion']>, 'Organization'> } &
  { Repository: UU<NN<GetSimpleFollowSuggestionsQuery['followSuggestion']>, 'Repository'> } &
  { Query: GetSimpleFollowSuggestionsQuery }

  export type TypeUsagesForSubmitRepositoryMutation = { Entry: SubmitRepositoryMutation['submitRepository'] } &
  { Mutation: SubmitRepositoryMutation }

  export type TypeUsagesForGetRepositoryContributorsQuery = { Repository: GetRepositoryContributorsQuery['repository'] } &
  { Actor: NN<GetRepositoryContributorsQuery['repository']>['owner'] } &
  { User: UU<NN<GetRepositoryContributorsQuery['repository']>['owner'], 'User'> } &
  { Actor: NN<NN<GetRepositoryContributorsQuery['repository']>['contributors']>[0] } &
  { User: UU<NN<NN<NN<GetRepositoryContributorsQuery['repository']>['contributors']>[0]>, 'User'> } &
  { Organization: UU<NN<NN<NN<GetRepositoryContributorsQuery['repository']>['contributors']>[0]>, 'Organization'> } &
  { Query: GetRepositoryContributorsQuery }

  export type TypeUsagesForSubmitCommentMutation = { Comment: SubmitCommentMutation['submitComment'] } &
  { Mutation: SubmitCommentMutation } &
  TypeUsagesForCommentsPageCommentFragment

  export type TypeUsagesForVoteMutation = { Entry: VoteMutation['vote'] } &
  { Vote: NN<VoteMutation['vote']>['vote'] } &
  { Mutation: VoteMutation }

export interface OperationsMap {
  'CommentsPageComment': { operationType: CommentsPageCommentFragment, variablesType: {}, typeUsages: TypeUsagesForCommentsPageCommentFragment, kind: 'fragment' }
  'FeedEntry': { operationType: FeedEntryFragment, variablesType: {}, typeUsages: TypeUsagesForFeedEntryFragment, kind: 'fragment' }
  'RepoInfo': { operationType: RepoInfoFragment, variablesType: {}, typeUsages: TypeUsagesForRepoInfoFragment, kind: 'fragment' }
  'VoteButtons': { operationType: VoteButtonsFragment, variablesType: {}, typeUsages: TypeUsagesForVoteButtonsFragment, kind: 'fragment' }
  'onCommentAdded': { operationType: OnCommentAddedSubscription, variablesType: OnCommentAddedSubscriptionVariables, typeUsages: TypeUsagesForOnCommentAddedSubscription, kind: 'subscription' }
  'Comment': { operationType: CommentQuery, variablesType: CommentQueryVariables, typeUsages: TypeUsagesForCommentQuery, kind: 'query' }
  'CurrentUserForProfile': { operationType: CurrentUserForProfileQuery, variablesType: CurrentUserForProfileQueryVariables, typeUsages: TypeUsagesForCurrentUserForProfileQuery, kind: 'query' }
  'Feed': { operationType: FeedQuery, variablesType: FeedQueryVariables, typeUsages: TypeUsagesForFeedQuery, kind: 'query' }
  'getExtendedFollowSuggestions': { operationType: GetExtendedFollowSuggestionsQuery, variablesType: GetExtendedFollowSuggestionsQueryVariables, typeUsages: TypeUsagesForGetExtendedFollowSuggestionsQuery, kind: 'query' }
  'getSimpleFollowSuggestions': { operationType: GetSimpleFollowSuggestionsQuery, variablesType: GetSimpleFollowSuggestionsQueryVariables, typeUsages: TypeUsagesForGetSimpleFollowSuggestionsQuery, kind: 'query' }
  'submitRepository': { operationType: SubmitRepositoryMutation, variablesType: SubmitRepositoryMutationVariables, typeUsages: TypeUsagesForSubmitRepositoryMutation, kind: 'mutation' }
  'getRepositoryContributors': { operationType: GetRepositoryContributorsQuery, variablesType: GetRepositoryContributorsQueryVariables, typeUsages: TypeUsagesForGetRepositoryContributorsQuery, kind: 'query' }
  'submitComment': { operationType: SubmitCommentMutation, variablesType: SubmitCommentMutationVariables, typeUsages: TypeUsagesForSubmitCommentMutation, kind: 'mutation' }
  'vote': { operationType: VoteMutation, variablesType: VoteMutationVariables, typeUsages: TypeUsagesForVoteMutation, kind: 'mutation' }
}

export interface TypesMap {
  operations: OperationsMap
  fieldArgsUsages: FieldArgsUsagesMap
  allOutputTypes: AllOutputTypes
  allScalarTypes: AllScalarTypes
}

export const documentsMap = {
  'CommentsPageComment': { document: CommentsPageCommentFragmentDoc, kind: 'fragment' as const },
  'FeedEntry': { document: FeedEntryFragmentDoc, kind: 'fragment' as const },
  'RepoInfo': { document: RepoInfoFragmentDoc, kind: 'fragment' as const },
  'VoteButtons': { document: VoteButtonsFragmentDoc, kind: 'fragment' as const },
  'onCommentAdded': { document: OnCommentAddedDocument, kind: 'subscription' as const },
  'Comment': { document: CommentDocument, kind: 'query' as const },
  'CurrentUserForProfile': { document: CurrentUserForProfileDocument, kind: 'query' as const },
  'Feed': { document: FeedDocument, kind: 'query' as const },
  'getExtendedFollowSuggestions': { document: GetExtendedFollowSuggestionsDocument, kind: 'query' as const },
  'getSimpleFollowSuggestions': { document: GetSimpleFollowSuggestionsDocument, kind: 'query' as const },
  'submitRepository': { document: SubmitRepositoryDocument, kind: 'mutation' as const },
  'getRepositoryContributors': { document: GetRepositoryContributorsDocument, kind: 'query' as const },
  'submitComment': { document: SubmitCommentDocument, kind: 'mutation' as const },
  'vote': { document: VoteDocument, kind: 'mutation' as const }
}",
  "prepend": Array [
    "type NonNullable<T> = T extends null | undefined ? never : T",
    "type NN<T> = NonNullable<T>",
    "type UnwrapUnion<T, K> = T extends { __typename: K } ? T : never",
    "type UU<T, K> = UnwrapUnion<T, K>",
  ],
}
`;
